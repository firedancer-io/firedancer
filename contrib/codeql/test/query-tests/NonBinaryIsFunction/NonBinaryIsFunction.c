// Add missing typedef and define
#define FD_EXECUTOR_INSTR_ERR_MISSING_ACC (-33) /* An account required by the instruction is missing */
#define FD_BANK_FLAGS_DEAD (0x00000008UL)       /* Dead.  We stopped replaying it before we could finish it (e.g. invalid block or pruned minority fork). */

typedef struct
{
    int acct_cnt;
    struct
    {
        int is_signer;
    } accounts[10];
} fd_instr_info_t;

typedef struct
{
    int flags;
} fd_bank_t;

int fd_instr_acc_is_signer_idx(fd_instr_info_t const *instr,
                               short idx)
{
    if (FD_UNLIKELY(idx >= instr->acct_cnt))
    {
        return FD_EXECUTOR_INSTR_ERR_MISSING_ACC; // $ Alert (returns -33 which is not 0 or 1)
    }

    return !!(instr->accounts[idx].is_signer);
}

static inline int
fd_banks_is_bank_dead(fd_bank_t *bank)
{
    return bank->flags & FD_BANK_FLAGS_DEAD; // $ Alert (returns value in [0, 8])
}

// inspired by the function, but heavily simplified
int fd_bn254_pairing_is_one_syscall(unsigned long in_sz)
{
    if (in_sz != 128UL)
    {
        return -1; // No alert
    }
    return 0;
}