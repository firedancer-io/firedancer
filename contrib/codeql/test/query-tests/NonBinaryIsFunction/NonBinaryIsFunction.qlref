query: NonBinaryIsFunction.ql
postprocess: InlineExpectationsTestQuery.ql