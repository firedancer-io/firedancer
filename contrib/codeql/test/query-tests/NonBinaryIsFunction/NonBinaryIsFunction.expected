| NonBinaryIsFunction.c:24:9:24:49 | return ... | The function $@ is named like an `is` function but returns a value with bounds [-33, -33]. | NonBinaryIsFunction.c:19:5:19:30 | fd_instr_acc_is_signer_idx | fd_instr_acc_is_signer_idx |
| NonBinaryIsFunction.c:33:5:33:44 | return ... | The function $@ is named like an `is` function but returns a value with bounds [0, 8]. | NonBinaryIsFunction.c:31:1:31:21 | fd_banks_is_bank_dead | fd_banks_is_bank_dead |
